README
snake4 - Fruit-eating snake game
================================

This program is my fourth implementation of the snake game, written
for Unix and X11.  Previous versions were written for MS-DOS.

Move the snake around the screen and eat food.  The snake grows in
length as you eat.  Avoid hitting the fence, eating lethal mushrooms
and rotten food, and biting your own tail.

Now and then a slim-pill will bounce around.  If you eat this, you
will get shorter, and you will gain bonus points for every length you
loose.  The bonus is also given when a round is over.

When food starts blinking, it's about to rot.  Rotten food is
poisonous, and thus uneatable.  As long as it blinks, it can still be
eaten.

Oh, yes - I almost forgot: You may be visited by the evil headbanger.
Take my advice and run away from him before he hits you in the head!
He won't stay long.

Thx to brother Reidar for thinking up the slim-pill.

Choose among the following keys to move the snake around:

        ------+---+---+---+------------
	Up    | A | T | K | Up arrow
	Down  | Z | V | J | Down arrow
	Left  | N | F | H | Left arrow
	Right | M | G | L | Right arrow
        ------+---+---+---+------------

Use `space', `pause' or `p' to pause and unpause, and `esc' or `q' to
quit.

snake4 features a site-wide highscore file.  The file contains the top
200 users.  Each user may appear once only in the highscore table.


Author
------

The program is written by

        Sverre H. Huseby        shh@thathost.com
        Lofthusvn. 11 B         http://shh.thathost.com/
        N-0587 Oslo
        Norway


License
-------

This program is released under the Artistic License:

  http://www.opensource.org/licenses/artistic-license.html

Comments (even as simple as "I use your program") are very welcome.
If you insist on paying something, please donate some money to an
organization that strives to make the world a better place for
everyone.

I don't like bugs, so please help me removing them by reporting
whatever you find!



ChangeLog
2014-08-14  Alexandre Dantas  <eu@alexdantas.net>

	* Release: 1.0.14

	* Makefile fix after renaming README to README.md

2014-08-13  Sverre H. Huseby  <shh@thathost.com>

	* Release: 1.0.13

	* Code cleanup by Alexandre Dantas

2002-03-02  Sverre H. Huseby  <shh@thathost.com>

	* Release: 1.0.12

	* All: Updated contact information.

1998-07-05  Sverre H. Huseby  <shh@thathost.com>

	* Release: 1.0.11

	* Updated author information, and changed license to Artistic
	License, to please sunsite.

	* pixmaps/instructions: removed old mail address.

Wed Aug 13 22:39:30 1997  Sverre H. Huseby  <s.h.huseby@usit.uio.no>

	* Release: 1.0.10

	* score.c (scoreWriteScoreLines): Makes sure no empty fields are
	written out, as strtok doesn't like empty tokens.

Mon Aug 11 21:30:35 1997  Sverre H. Huseby  <s.h.huseby@usit.uio.no>

	* Release: 1.0.9

	* Makefile: Added possibility to make a binary distribution.

Wed Mar 19 20:11:25 1997  Sverre H. Huseby  <s.h.huseby@usit.uio.no>

	* Version: 1.0.8

	* win.c (winParseOptions): Added a terminating NULL to the
	resources.

Mon Nov 11 22:04:24 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.7

	* score.c: Use 3 positions for position number on score list.

Tue Oct 15 11:32:10 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.6

	* pixmaps/instructions*: Included score values in the instructions.

Fri Oct 11 11:15:15 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.5

	* score.c (scoreGetThisPlayerIndex): New function.

	* board.c (boardShowHighscores): No longer shows the top 20, but
	rather the top 10, plus a few centered around the current player.

Fri Oct 11 11:15:15 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.4

	* snake.c: Added idling timeout for start of rounds. The game is
	over if the user waits too long (a couple of minutes) before
	moving the snake.

Fri Oct 11 10:31:55 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.3

	* Increased number of entries in the highscore file, and made a
	constant for how many scores may be showed inline.

Thu Oct 10 13:41:53 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.2

        * game.c (gameInit): Fixed bug. Random seed was initialized using
        clock(), but clock returns number of ticks since the last call to
        clock(), so it was always zero.

Wed Oct  9 17:48:49 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.1

	* snake4.c (main): Now shows highscores if called as snake4scores.

	* Makefile: At installation, creates a symlink called snake4scores
	to snake4.

Tue Oct  8 06:56:01 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 1.0.0 (uploaded to sunsite.unc.edu)

	* win.c (winTogglePause): Bugfix: Added test for
	gameEndOfRound. Problems occurred if the player paused the game
	after the snake died.

	* board.c (boardHandleEvent): Moved test for pausing legal to
	winTogglePause in win.c

	* suid.c: Added this file and suid.h. Fixes SUID-problems on
	SunOS.

	* Makefile: Creates snake4.score from snake4.score.ifi by keeping
	sverrehu only.

	* Translated TODO to English.

Sun Oct  6 20:38:50 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 0.9.5

	* snake4.c (version): Added info on who compiled the program, and
	when.

Sun Oct  6 17:03:50 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 0.9.4

	* board.c (boardHandleEvent): Bugfix: Expose events would both
	show instructions and repaint game objects.

Sun Oct  6 15:05:19 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version: 0.9.3

	* slimpill.c: Slimpill comes less often.

	* headbanger.c: Score for each move of the headbanger, and extra
	bonus when it leaves.

	* fruit.c (fruitHandleTick): At least two fruits available.

Sat Oct  5 11:19:52 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Version 0.9.2

	* snake4.6: Added a man-page.

	* score.c, snake4.c: Added possibility to merge scores from
 	separate highscore file.

Fri Oct  4 22:14:57 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Bumped version up to 0.9.1 - soon finished!

Tue Sep 24 19:38:47 1996  Sverre H. Huseby  <sverrehu@ifi.uio.no>

	* Initial version. This program was first implemented in assembly
	for MS-DOS by the same author.
	



win.c
/* $Id: win.c,v 1.6 2002/03/02 21:10:55 sverrehu Exp $ */
/**************************************************************************
 *
 *  FILE            win.c
 *  MODULE OF       snake4 - game of snake eating fruit
 *
 *  DESCRIPTION     
 *
 *  WRITTEN BY      Sverre H. Huseby <shh@thathost.com>
 *
 **************************************************************************/

#include <stdlib.h>

#include <X11/Xlib.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Core.h>
#include <X11/Xaw/Command.h>
#include <X11/Xaw/Label.h>
#include <X11/Xaw/Form.h>

#include <shhmsg.h>

#include "game.h"
#include "board.h"
#include "snake.h"
#include "score.h"
#include "win.h"

/**************************************************************************
 *                                                                        *
 *                       P R I V A T E    D A T A                         *
 *                                                                        *
 **************************************************************************/

#define TICK_FREQ 30

static int winSleepDone;
static XtIntervalId timerId;
static Widget chooseLevelWidget, startWidget[MAX_LEVEL - MIN_LEVEL + 1];
static Widget levelWidget, roundWidget, scoreWidget, highscoreWidget;



/**************************************************************************
 *                                                                        *
 *                        P U B L I C    D A T A                          *
 *                                                                        *
 **************************************************************************/

Display *winDisplay;

XtAppContext winAppContext;
Widget winTopLevel;
Widget winMainWindow;


/**************************************************************************
 *                                                                        *
 *                   P R I V A T E    F U N C T I O N S                   *
 *                                                                        *
 **************************************************************************/

static void
winGetAndHandleEvent(void)
{
    XEvent e;

    XtAppNextEvent(winAppContext, &e);
    if (e.xany.window == boardWin || e.type == KeyPress)
	boardHandleEvent(&e);
    else if (e.type == FocusIn) {
	XAutoRepeatOff(winDisplay);  /* for users holding the keys down */
    } else if (e.type == FocusOut) {
	XAutoRepeatOn(winDisplay);
    } else
	XtDispatchEvent(&e);
}

static void
winSleepTimerProc(XtPointer foo, XtIntervalId *bar)
{
    XEvent e;

    winSleepDone = 1;
    /* wake up the waiting loop by sending a ClientMessage */
    e.type = ClientMessage;
    e.xclient.message_type = 0;
    e.xclient.format = 8;
    XSendEvent(winDisplay, XtWindowOfObject(winMainWindow), 0, 0, &e);
}

static void
winSleep(unsigned long ms)
{
    winSleepDone = 0;
    XtAppAddTimeOut(winAppContext, ms, winSleepTimerProc, 0);
    while (!winSleepDone && !gameExit)
	winGetAndHandleEvent();
}

static void timerProc(XtPointer, XtIntervalId *);

static void
installTimer(void)
{
    timerId = XtAppAddTimeOut(winAppContext, TICK_FREQ, timerProc, 0);
}

static void
removeTimer(void)
{
    XtRemoveTimeOut(timerId);
}

static void
timerProc(XtPointer foo, XtIntervalId *bar)
{
    installTimer();
    gameHandleTick();
    if (gameEndOfRound) {
	removeTimer();
	if (gameEndOfRound == 1)
	    winSleep(1000);
	gameFinishRound();
	if (gameRound > MAX_ROUNDS || gameEndOfRound > 1) {
	    int q;

	    gameFinishGame();

	    XtSetSensitive(chooseLevelWidget, 1);
	    for (q = 0; q < MAX_LEVEL - MIN_LEVEL + 1; q++)
		XtSetSensitive(startWidget[q], 1);
	    XtSetSensitive(levelWidget, 0);
	    XtSetSensitive(roundWidget, 0);
	    XtSetSensitive(scoreWidget, 0);
	    XtSetSensitive(highscoreWidget, 0);
	} else {
	    installTimer();
	    gameInitRound();
	}
    }
}

static void
callbackQuit(Widget w, XtPointer clientData, XtPointer callData)
{
    gameExit = 1;
}

static void
callbackStart(Widget w, XtPointer clientData, XtPointer callData)
{
    winChooseLevel((int) clientData);
}



/**************************************************************************
 *                                                                        *
 *                    P U B L I C    F U N C T I O N S                    *
 *                                                                        *
 **************************************************************************/

void
winParseOptions(int *argc, char *argv[])
{
    static String resources[] = {
	"*board.accelerators: <Key>:",
	"*background: black",
	"*foreground: white",
	"*borderColor: dim gray",
	"*Command.background: gray",
	"*Command.foreground: black",
	"*score.foreground: yellow",
	"*font: 9x15bold",
	NULL
    };

    winTopLevel = XtVaAppInitialize(&winAppContext, "Snake4",
                                    NULL, 0, argc, argv, resources, NULL);
    winDisplay = XtDisplay(winTopLevel);
}

void
winInit(void)
{
    int q;
    char num[10];
    Widget boardWidget;
    Widget quitWidget;

    winMainWindow
	= XtVaCreateManagedWidget("mainWindow",
				  formWidgetClass, winTopLevel,
				  
				  NULL);
    boardWidget
	= XtVaCreateManagedWidget("board",
				  coreWidgetClass, winMainWindow,
				  XtNwidth, boardWidth,
				  XtNheight, boardHeight,
				  XtNborderWidth, 0,
				  NULL);

    chooseLevelWidget
	= XtVaCreateManagedWidget("chooseLevel",
				  labelWidgetClass, winMainWindow,
				  XtNlabel, "Choose level:",
				  XtNfromVert, boardWidget,
				  NULL);
    for (q = 0; q < MAX_LEVEL - MIN_LEVEL + 1; q++) {
	sprintf(num, "%d", q + MIN_LEVEL);
	startWidget[q]
	    = XtVaCreateManagedWidget(num,
				      commandWidgetClass, winMainWindow,
				      XtNfromHoriz, 
				        q ? startWidget[q - 1]
				          : chooseLevelWidget,
				      XtNfromVert, boardWidget,
				      NULL);
	XtAddCallback(startWidget[q], XtNcallback, callbackStart,
		      (XtPointer) (q + MIN_LEVEL));
    }

    levelWidget
	= XtVaCreateManagedWidget("level",
				  labelWidgetClass, winMainWindow,
				  XtNlabel, "Level: 0",
				  XtNfromVert, boardWidget,
				  XtNfromHoriz,
				    startWidget[MAX_LEVEL - MIN_LEVEL],
				  XtNsensitive, 0,
				  NULL);
    roundWidget
	= XtVaCreateManagedWidget("round",
				  labelWidgetClass, winMainWindow,
				  XtNlabel, "Lives: 0",
				  XtNfromVert, boardWidget,
				  XtNfromHoriz, levelWidget,
				  XtNsensitive, 0,
				  NULL);
    scoreWidget
	= XtVaCreateManagedWidget("score",
				  labelWidgetClass, winMainWindow,
				  XtNlabel, "Score:      0",
				  XtNfromVert, boardWidget,
				  XtNfromHoriz, roundWidget,
				  XtNsensitive, 0,
				  NULL);
    highscoreWidget
	= XtVaCreateManagedWidget("highscore",
				  labelWidgetClass, winMainWindow,
				  XtNlabel, "High:      0",
				  XtNfromVert, boardWidget,
				  XtNfromHoriz, scoreWidget,
				  XtNsensitive, 0,
				  NULL);
    quitWidget
	= XtVaCreateManagedWidget("quit",
				  commandWidgetClass, winMainWindow,
				  XtNlabel, "Quit",
				  XtNfromVert, boardWidget,
				  XtNfromHoriz, highscoreWidget,
				  NULL);
    XtAddCallback(quitWidget, XtNcallback, callbackQuit, 0);

    XtRealizeWidget(winTopLevel);
    XtInstallAccelerators(winTopLevel, boardWidget);

    boardWin = XtWindowOfObject(boardWidget);
}

void
winFinish(void)
{
}

void
winMainLoop(void)
{
    while (!gameExit)
	winGetAndHandleEvent();
    if (gameRunning)
	scoreFinishGame();
    XAutoRepeatOn(winDisplay);
    winFlush();
}

void
winFlush(void)
{
    XFlush(winDisplay);
}

void
winTogglePause(void)
{
    if (!gameRunning || gameEndOfRound)
	return;
    gamePause ^= 1;
    if (!gamePause)
	snakeFlushKeys();
    if (gamePause)
	removeTimer();
    else
	installTimer();
}

void
winChooseLevel(int level)
{
    int q;

    if (gameRunning || level < MIN_LEVEL || level > MAX_LEVEL)
	return;
    XtSetSensitive(chooseLevelWidget, 0);
    for (q = 0; q < MAX_LEVEL - MIN_LEVEL + 1; q++)
	XtSetSensitive(startWidget[q], 0);
    XtSetSensitive(levelWidget, 1);
    XtSetSensitive(roundWidget, 1);
    XtSetSensitive(scoreWidget, 1);
    XtSetSensitive(highscoreWidget, 1);

    gameSetLevel(level);
    gameInitGame();
    gameInitRound();
    installTimer();
}

void
winShowLevel(void)
{
    char text[40];

    sprintf(text, "Level: %d", gameLevel);
    XtVaSetValues(levelWidget, XtNlabel, text, NULL);
}

void
winShowRound(void)
{
    char text[40];

    sprintf(text, "Lives: %d", MAX_ROUNDS - gameRound + 1);
    XtVaSetValues(roundWidget, XtNlabel, text, NULL);
}

void
winShowScore(void)
{
    char text[40];

    sprintf(text, "Score:%7ld", gameScore);
    XtVaSetValues(scoreWidget, XtNlabel, text, NULL);
}

void
winShowHighscore(void)
{
    char text[40];

    sprintf(text, "High:%7ld", gameHighscore);
    XtVaSetValues(highscoreWidget, XtNlabel, text, NULL);
}
